File: business/src/main/java/com/pij/lottieshow/interactor/CompoundSerializer.java
package com.pij.lottieshow.interactor;

import android.support.annotation.NonNull;

import java.net.URI;
import java.util.NoSuchElementException;

import javax.inject.Inject;

import rx.Observable;
import rx.Single;

import static rx.Observable.empty;
import static rx.Observable.error;
import static rx.Observable.from;

public class CompoundSerializer implements Serializer {

    private final Iterable<Serializer> serializers;

    @Inject
    @SuppressWarnings("WeakerAccess")
    public CompoundSerializer(@NonNull Iterable<Serializer> sources) {
        this.serializers = sources;
    }

    @Override
    public Single<String> open(URI input) {
        // emits all known formats
        Observable<String> contents = from(serializers).concatMap(serializer -> serializer.open(input)
                                                                                          .toObservable()
                                                                                          .onErrorResumeNext
                                                                                                  (this::absorbUnknownFile));
        return contents.take(1).toSingle().onErrorResumeNext(e -> transformNoSerializerFound(e, input));
    }

    private Single<? extends String> transformNoSerializerFound(Throwable e, URI file) {
        return (e instanceof NoSuchElementException) ? Single.error(new UnknownFileException(
                "Could not find a Serializer for " + file + ".",
                e)) : Single.error(e);
    }

    private Observable<String> absorbUnknownFile(Throwable e) {
        return (e instanceof UnknownFileException) ? empty() : error(e);
    }

}

File: business/src/main/java/com/pij/lottieshow/interactor/CompoundSerializer.kt
package com.pij.lottieshow.interactor

import rx.Observable
import rx.Observable.combineLatest
import rx.Observable.empty
import rx.Observable.error
import rx.Observable.from
import rx.Observable.just
import rx.Single
import java.net.URI
import javax.inject.Inject

class CompoundSerializer @Inject constructor(private val serializers: Iterable<Serializer>) : Serializer {

    override fun open(input: URI): Single<String> {
        // emits all known formats
        val content = combineLatest(from(serializers), just(input), this::open).concatMap { it }
        return content.take(1).toSingle().onErrorResumeNext { transformNoSerializerFound(it, input) }
    }

    private fun open(opener: Serializer, input: URI): Observable<String> {
        return opener.open(input).toObservable().onErrorResumeNext { absorbUnknownFile(it) }
    }

    private fun transformNoSerializerFound(e: Throwable, file: URI): Single<String> {
        val result = if (e is NoSuchElementException) {
            UnknownFileException("Could not find a Serializer for $file.", e)
        } else {
            e
        }
        return Single.error(result)
    }

    private fun absorbUnknownFile(e: Throwable): Observable<String> {
        return if (e is UnknownFileException) empty() else error(e)
    }

}
